src/sparse_blas/backends/cusparse/operations/cusparse_spmm.cpp

```diff
@@ -0,0 +1,296 @@
+*
+**************************************************************************/
+
+#include "oneapi/mkl/sparse_blas/detail/cusparse/onemkl_sparse_blas_cusparse.hpp"
+
+#include "sparse_blas/backends/cusparse/cusparse_error.hpp"
+#include "sparse_blas/backends/cusparse/cusparse_helper.hpp"
+#include "sparse_blas/backends/cusparse/cusparse_task.hpp"
+#include "sparse_blas/backends/cusparse/cusparse_handles.hpp"
+#include "sparse_blas/common_op_verification.hpp"
+#include "sparse_blas/macros.hpp"
+#include "sparse_blas/matrix_view_comparison.hpp"
+#include "sparse_blas/sycl_helper.hpp"
+
+namespace oneapi::mkl::sparse {
+
+// Complete the definition of the incomplete type
+struct spmm_descr {
+    detail::generic_container workspace;
+    std::size_t temp_buffer_size = 0;
+    bool buffer_size_called = false;
+    bool optimized_called = false;
+    oneapi::mkl::transpose last_optimized_opA;
+    oneapi::mkl::transpose last_optimized_opB;
+    oneapi::mkl::sparse::matrix_view last_optimized_A_view;
+    oneapi::mkl::sparse::matrix_handle_t last_optimized_A_handle;
+    oneapi::mkl::sparse::dense_matrix_handle_t last_optimized_B_handle;
+    oneapi::mkl::sparse::dense_matrix_handle_t last_optimized_C_handle;
+    oneapi::mkl::sparse::spmm_alg last_optimized_alg;
+};
+
+} // namespace oneapi::mkl::sparse
+
+namespace oneapi::mkl::sparse::cusparse {
+
+void init_spmm_descr(sycl::queue& /*queue*/, spmm_descr_t* p_spmm_descr) {
+    *p_spmm_descr = new spmm_descr();
+}
+
+sycl::event release_spmm_descr(sycl::queue& queue, spmm_descr_t spmm_descr,
+                               const std::vector<sycl::event>& dependencies) {
+    return detail::submit_release(queue, spmm_descr, dependencies);
+}
+
+inline auto get_cuda_spmm_alg(spmm_alg alg) {
+    switch (alg) {
+        case spmm_alg::coo_alg1: return CUSPARSE_SPMM_COO_ALG1;
+        case spmm_alg::coo_alg2: return CUSPARSE_SPMM_COO_ALG2;
+        case spmm_alg::coo_alg3: return CUSPARSE_SPMM_COO_ALG3;
+        case spmm_alg::coo_alg4: return CUSPARSE_SPMM_COO_ALG4;
+        case spmm_alg::csr_alg1: return CUSPARSE_SPMM_CSR_ALG1;
+        case spmm_alg::csr_alg2: return CUSPARSE_SPMM_CSR_ALG2;
+        case spmm_alg::csr_alg3: return CUSPARSE_SPMM_CSR_ALG3;
+        default: return CUSPARSE_SPMM_ALG_DEFAULT;
+    }
+}
+
+inline void fallback_alg_if_needed(oneapi::mkl::sparse::spmm_alg& alg, oneapi::mkl::transpose opA,
+                                   oneapi::mkl::transpose opB) {
+    if (alg == oneapi::mkl::sparse::spmm_alg::csr_alg3 &&
+        (opA != oneapi::mkl::transpose::nontrans || opB == oneapi::mkl::transpose::conjtrans)) {
+        // Avoid warnings printed on std::cerr
+        alg = oneapi::mkl::sparse::spmm_alg::default_alg;
+    }
+}
+
+void spmm_buffer_size(sycl::queue& queue, oneapi::mkl::transpose opA, oneapi::mkl::transpose opB,
+                      const void* alpha, oneapi::mkl::sparse::matrix_view A_view,
+                      oneapi::mkl::sparse::matrix_handle_t A_handle,
```

Comment:
Because this is under the `oneapi::mkl::sparse::cusparse` namespace, I believe we can remove the `oneapi::mkl::sparse` namespace used many times here. If nothing, at least remove the `oneapi::mkl::` part. It is a bit distracting to see that, and this file is short enough for us to know what namespace we are dealing with.

Reply:
True, I have removed some redundant namepace for cuSPARSE in https://github.com/oneapi-src/oneMKL/pull/527/commits/84565a9a7725be6efb03ef5bf378b36997d0bb18.
Note I have also removed some for the MKLCPU and MKLGPU backends in a separate commit https://github.com/oneapi-src/oneMKL/pull/527/commits/44dc73d8f8f4fdbef79bf06f140c0cc7d4349b4b. This PR already impacted the MKL backends but let me know if you prefer to split this separately.